crawl-ref/source/player.cc

```diff
@@ -4745,11 +4745,24 @@ void dec_channel_player(int delay)
 bool invis_allowed(bool quiet, string *fail_reason)
 {
     string msg;
+    string reason = "";
     bool success = true;
 
     if (you.haloed() && you.halo_radius() != -1)
     {
-        msg = "Your halo prevents invisibility.";
+        if (player_equip_unrand(UNRAND_EOS))
+            reason = "Your weapon shines too brightly";
```

Comment:
It feels unsatisfying to me to set this reason and then perhaps overwrite it later. In particular, this approach only works because you currently have exactly two halo sources. If there were more, you wouldn't be able to know which halo source had previously been set.

Better, I think, is to have non-overlapping `if` clauses, like this:

```
bool divine = (...);
bool weapon = (...);

if (divine && weapon)
  ...
else if (divine)
  ...
else if (weapon)
  ...
else
  die("haloed by an unknown source");
```